sota-implementations/post-training/grpo.py

```diff
@@ -0,0 +1,167 @@
+    answer_ids = tree_map(lambda string: hash(string), answer)
+    answer_ids = torch.tensor(answer_ids)
+    unique_qs = answer_ids.view(-1).unique()
+    trajectories["advantage"] = trajectories["next", "reward"] * 0
+    for u in unique_qs:
+        idx = answer_ids == u
+        rewards = trajectories[idx]["next", "reward"]
+        rewards = (rewards - rewards.mean()) / rewards.std().clamp(min=1e-4)
+        trajectories.set_at_("advantage", rewards, idx)
+    return trajectories
+
+
+if __name__ == "__main__":
+    args = parser.parse_args()
+    # Create env instance:
+    #  - Load the gsm8k dataset
+    dataset = load_dataset(args.dataset, "main")
+    train_dataset = dataset["train"]
+
+    def collate_fn(batch):
+        batch = torch.stack([TensorDict.from_dict(_batch) for _batch in batch])
+        batch.rename_key_("question", "text")
+        return batch
+
+    # LLM
+    tokenizer = AutoTokenizer.from_pretrained("gpt2")
+    # inference_model = GPT2LMHeadModel(GPT2Config())
+    inference_model = LLM("gpt2")
+    tokenizer.pad_token = tokenizer.eos_token
+    tokenizer.padding_side = "left"
+
+    # Env
+    dataloader = DataLoader(  # noqa: TOR401
+        train_dataset, batch_size=args.batch_size, shuffle=True, collate_fn=collate_fn
+    )
+    env = LLMEnv.from_dataloader(
+        dataloader=dataloader,
+        tokenizer=tokenizer,
+        str2str=True,
+        batch_size=(args.batch_size * args.repeats,),
+        repeats=args.repeats,
+    )
+    for i, trsf in enumerate(env.transform):
+        if isinstance(trsf, DataLoadingPrimer):
+            env.insert_transform(i, PrepareQuestion())
+            break
+
+    # Finally, we want the env to stop after the first step
+    env.append_transform(StepCounter(max_steps=1))
+
+    policy = from_vllm(
+        inference_model,
+        tokenizer=tokenizer,
+        from_text=False,
+        generate=True,
+        return_log_probs=True,
+    )
+
+    # Reward transform
+    env.append_transform(ShapedCorrectnessReward(tokenizer=tokenizer))
+
+    # Ref model
+    ref_model = GPT2LMHeadModel.from_pretrained("gpt2")
+    TensorDict.from_module(ref_model).data.to_module(ref_model)
+    ref_model = from_hf_transformers(
+        ref_model,
+        tokenizer=tokenizer,
+        from_text=False,
+        generate=False,
+        return_log_probs=True,
+    )
+    env.append_transform(
+        KLRewardTransform(actor=ref_model, coef=0.1, log_prob_key="log_probs")
+    )
+
+    # replay buffer
+    rb = ReplayBuffer(
+        storage=LazyStackStorage(args.steps_per_batch),
+        sampler=SamplerWithoutReplacement(),
+        batch_size=args.optim_batch_size,
+    )
+
+    # Collector
+    train_model = GPT2LMHeadModel.from_pretrained("gpt2")
+    collector = SyncDataCollector(
+        env,
+        policy,
+        frames_per_batch=args.steps_per_batch,
+        total_frames=1_000_000,
+        local_weights_updater=HF2vLLMLocalWeightUpdater(
```

Comment:
I'm curious to see how this would look when train_model is sharded :b

Reply:
Yeah there is still some work to be done!
IIUC your implementation you do a `full_tensor()` then you send that to the vllm weights right?